### ISSUE TITLE
[CSV-Importer] Improve mapping list

### ISSUE THREAD

**Issue by MacPaille** [firefly-iii/firefly-iii#3628]
**Description**
While using the CSV-importer I want to map all the entries I get from my bank to the nice friendly names I have set in Firefly.
The problem I'm facing is I have a lot of records and sometimes I have the same name for both expense account and revenue account (for example a friend I have send money but sometimes he also reimburse me, or a shop I buy stuff but sometimes I get a refund).
The dropdown list contains the title before starting the list of each account type, but when you are in the middle somewhere you don't know which account type you are in (I'm typing on the keyboard to go faster).

Second part I know I can use rules to map an account to a specific category but for now I'm still not doing it for all my accounts/categories. Also I don't have a category in my bank CSV I could use to do a mapping. It would be nice to have a dedicated field to add mappings of opposing account to specific categories that would be kept in the configuration file so the more we use the importer the more precise it will be

**Solution**
I think it may be easier to have the account type in front on each name like:
```
Expense accounts
  Expense accounts > Shop1
  Expense accounts > Shop2
  Expense accounts > Shop3
Revenue accounts
  Revenue accounts > Shop 1
  Revenue accounts > Shop 2
```
Instead of the current way:
```
Expense accounts
  Shop1
  Shop2
  Shop3
Revenue accounts
  Shop 1
  Shop 2
```
With such small list it's fine to use but I have 1600+ expense accounts and 300+ revenue accounts in mine :)

For the second part it would be to add a tickbox next to mapping to say map to categories where you will be able to choose a category for each account on the next page.

**What are alternatives?**
Keep the current system which works fine but is a bit harder to map entries when you have a long list

**Additional context**
Super big thanks for your work that is still amazing! I had hard time to record all my transactions manually (I have around a year back to fill in) but now I saw my bank provides me CSV export this will be super quickly solved 👯

**Comment by JC5** [comment]
You're hitting the limits of what the CSV importer can do, I'm afraid. Especially for the drop down lists. Wouldn't it be an idea to start doing the mapping in the JSON configuration directly?

To map accounts to categories, you could choose to copy the column with the account name, and use the second column to do the category mapping.

**Comment by MacPaille** [comment]
Yes I will check to do it directly in the json or anyway I'll pay attention on which account I select. Anyway the more I import (per batch of 10/20 lines) the faster it will go as I always keep the new config file as source for the next import :) Super good feature BTW

For the mapping of categories I'll continue with the rules then, don't want to change anything in the source CSV (I could do it but I would feel it's a waste of time haha).

Thanks for your quick replies again!

**Comment by MacPaille** [comment]
I repon it because I may have a suggestion on the mapping of opposing accounts and even mapping rules in general.
At the moment the config file generated contains only the mapping you have set for the import.

I would be nice to keep all the mapping list of the original file so the more you import and save it, the more precise it gets as you are mapping everything small batch at a time.
Mapping is much easier in the web interface than in the json file itself as you have to look the ID of each account one by one to create the entries :(

**Comment by JC5** [comment]
Alright I pushed code to do this to the develop branch, which is also available [on Docker](https://hub.docker.com/repository/docker/fireflyiii/csv-importer/tags?page=1&name=dev).

It should work, expanding your config file after every import instead of overwriting old values. It's a bit of a happy flow thing right now, so I'm curious to hear if you run into issues.

**Comment by MacPaille** [comment]
Cool I just deployed develop version of csv-importer. Will do some test next time I import my transactions.

At the moment my mapping list contains 403 entries and I use a bash script to sort it out (removing duplicate lines) and pushing this value to my configuration files for each bank account :)

**Comment by MacPaille** [comment]
Hey,

I checked yesterday with the DEV branch (which was version 2.2.2) and I did not saw any change in the mapping list. What was expected exactly? 😄

**Comment by JC5** [comment]
The idea is/was that previously mapped entries in your config.json would still be in the "new" json file (offered to download after the import). Even when you didn't use them. 

I just realised I never pushed the code back when I said I would though. :D 

It should be in 2.2.3 though, let me know.

**Comment by MacPaille** [comment]
Haha then I understand :) Will it work with both uploaded json and the ones present in the /configurations of the csv-importer?
I'll validate it next time then :)

**Comment by JC5** [comment]
Not yet, only uploaded configs. The local dir is treated read only atm.

**Comment by MacPaille** [comment]
I totally understand this folder to be kept as RO and it was not my idea (even if it's tempting :) ) What I mean it to propose to download the file like now but based on the one used as input from the configurations directory. So you can download the updated one and put it back in the same folder to be used next time.

In the end the best solution would be to have a seperated mapping list from the config files. It's normal each account has its own specifics regarding the the configs and which fields is used for the mappings, but the actual mapping list is normally common to each of your own accounts.

That's what my current bash script is doing. I kept my mapping list (419 entries now) in a dedicated file and my script does copy all this in each account config file.
